Rework `mypy` plugin documentation (#10650)
pydantic/pydantic: docs/integrations/mypy.md

With the pydantic mypy plugin, you can fearlessly refactor your models knowing mypy will catch any mistakes
if your field names or types change.

There are other benefits too! See below for more details.
Note that mypy already supports some features without using the Pydantic plugin, such as synthesizing a `__init__`  
method for Pydantic models and dataclasses. See the [mypy plugin capabilities](#mypy-plugin-capabilities) for a list
of additional features.

> ## Using mypy without the plugin

Author: Removed the section as it does not provide any useful info imo, and `strict_optional` isn't really relevant as of today (it was before because [historically mypy allowed this](https://mypy-lang.blogspot.com/2022/11/mypy-0990-released.html))

Author: https://github.com/pydantic/pydantic/pull/10651

To enable the plugin, just add `pydantic.mypy` to the list of plugins in your
[mypy config file](https://mypy.readthedocs.io/en/latest/config_file.html):

=== "`mypy.ini`"

    ```ini
    [mypy]
    plugins = pydantic.mypy
    ```

=== "`pyproject.toml`"

    ```toml
    [tool.mypy]
    plugins = ['pydantic.mypy']
    ```

!!! note
    If you're using `pydantic.v1` models, you'll need to add `pydantic.v1.mypy` to your list of plugins.

See the [plugin configuration](#configuring-the-plugin) for more details.

## Supported mypy versions

Pydantic supports the mypy versions released less than 6 months ago. Older versions may still work with the plugin

> but won't be tested. The list of released mypy versions can be found [here](https://mypy-lang.org/news.html). Note
> that the support policy is subject to changes if this becomes too much of an issue on our end.

Reviewer: ```suggestion
but won't be tested. The list of released mypy versions can be found [here](https://mypy-lang.org/news.html). Note
that the version support policy is subject to change at discretion of contributors.
```

## Mypy plugin capabilities

### Generate a `__init__` signature for Pydantic models

* Any required fields that don't have dynamically-determined aliases will be included as required
  keyword arguments.
* If the [`populate_by_name`][pydantic.ConfigDict.populate_by_name] model configuration value is set to
  `True`, the generated signature will use the field names rather than aliases.
* The [`init_forbid_extra`](#init_forbid_extra) and [`init_typed`](#init_typed) plugin configuration

>   values can further fine-tune the synthesized `__init__` method.

Reviewer: I feel like the explicit specifications here were helpful - people often ask about `init_typed`...

Author: Then can just click on the `[init_typed](#init_typed)` reference, where I added a description for the setting (and same for others). Having things centralized avoids duplication, and having a heading makes it easier to link to when responding on github issues/discussions etc

To be clear, the explanation is not removed, just centralized in the configuration options section

### Generate a typed signature for `model_construct`

* The [`model_construct`][pydantic.BaseModel.model_construct] method is an alternative to model validation when input data is
  known to be valid and should not be parsed (see the [documentation](../concepts/models.md#creating-models-without-validation)).
  Because this method performs no runtime validation, static checking is important to detect errors.

### Support for frozen models

* If the [`frozen`][pydantic.ConfigDict.frozen] configuration is set to `True`, you will get
  an error if you try mutating a model field (see [faux immutability](../concepts/models.md#faux-immutability))

### Respect the type of the `Field`'s `default` and `default_factory`

* Field with both a `default` and a `default_factory` will result in an error during static checking.

> * The type of the `default` and `default_factory` value must be compatible with the one of the field.

Author: I'll note that we could/should follow the definition of [`dataclasses.field`](https://github.com/python/typeshed/blob/a2ee32ba5dced792abf0a771737248549dfcf32c/stdlib/dataclasses.pyi#L148-L149) so that this becomes supported without the plugin as well.

Author: https://github.com/pydantic/pydantic/pull/10651

  values can further fine-tune the synthesized `__init__` method.

### Generate a typed signature for `model_construct`

* The [`model_construct`][pydantic.BaseModel.model_construct] method is an alternative to model validation when input data is
  known to be valid and should not be parsed (see the [documentation](../concepts/models.md#creating-models-without-validation)).
  Because this method performs no runtime validation, static checking is important to detect errors.

### Support for frozen models

* If the [`frozen`][pydantic.ConfigDict.frozen] configuration is set to `True`, you will get
  an error if you try mutating a model field (see [faux immutability](../concepts/models.md#faux-immutability))

### Respect the type of the `Field`'s `default` and `default_factory`

* Field with both a `default` and a `default_factory` will result in an error during static checking.
* The type of the `default` and `default_factory` value must be compatible with the one of the field.

### Warn about the use of untyped fields
* You'll get a mypy error any time you assign a public attribute on a model without annotating its type

> * If your goal is to set a ClassVar, you should explicitly annotate the field using typing.ClassVar

Author: This line is stated in the linked usage error docs

## Optional Capabilities:
* While defining a field without an annotation will result in a [runtime error](../errors/usage_errors.md#model-field-missing-annotation),
  the plugin will also emit a type checking error.

### Prevent the use of required dynamic aliases

* If the [`warn_required_dynamic_aliases` **plugin setting**](#configuring-the-plugin) is set to `True`, you'll get a mypy
  error any time you use a dynamically-determined alias or alias generator on a model with
  `Config.populate_by_name=False`.
* This is important because if such aliases are present, mypy cannot properly type check calls to `__init__`.
  In this case, it will default to treating all arguments as optional.
See the documentation of the [`warn_required_dynamic_aliases`](#warn_required_dynamic_aliases) plugin configuration value.

> ## Enabling the Plugin

Author: Section moved above.

Reviewer: Good call.
